server/apps/microservices/src/processors/asset-uploaded.processor.ts

```diff
@@ -0,0 +1,67 @@
+import { InjectQueue, OnQueueActive, OnQueueCompleted, OnQueueWaiting, Process, Processor } from '@nestjs/bull';
+import { Job, Queue } from 'bull';
+import { AssetEntity, AssetType } from '@app/database/entities/asset.entity';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { randomUUID } from 'crypto';
+
+@Processor('asset-uploaded-queue')
+export class AssetUploadedProcessor {
+  constructor(
+    @InjectQueue('thumbnail-generator-queue')
+    private thumbnailGeneratorQueue: Queue,
+
+    @InjectQueue('metadata-extraction-queue')
+    private metadataExtractionQueue: Queue,
+
+    @InjectQueue('video-conversion-queue')
+    private videoConversionQueue: Queue,
+
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+  ) {}
+
+  /**
+   * Post processing uploaded asset to perform the following function if missing
+   * 1. Generate JPEG Thumbnail
+   * 2. Generate Webp Thumbnail <-> if JPEG thumbnail exist
+   * 3. EXIF extractor
+   * 4. Reverse Geocoding
+   *
+   * @param job asset-uploaded
+   */
+  @Process('asset-uploaded')
+  async processUploadedVideo(job: Job) {
```

Comment:
This is the starting point of the microservice that handles post-upload-related tasks. It will then put the additional jobs into the corresponding queues.

* Generate JPEG Thumbnail
* Generate Webp Thumbnail <-> if JPEG thumbnail exists
* EXIF extractor
* Reverse Geocoding
* Video Encoding
   
This is the second stage in the diagram.

![immich-queue-layour copy 2](https://user-images.githubusercontent.com/27055614/172532851-2416a252-91ef-4936-af28-3135d4860bea.png)